pkg/cmd/mysqlout-to-csv/convert.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+
+package main
+
+import (
+	"io"
+	"unicode/utf8"
+
+	"github.com/pkg/errors"
+)
+
+type dumpReader struct {
+	rows  int
+	width int
+
+	f func([]string) error
+
+	fieldSep rune // default \t
+	rowSep   rune // default \n
+
+	hasEscapeChar  bool
+	escapeChar     rune
+	hasEncloseChar bool
+	encloseChar    rune
+
+	bufSizeKb int
+}
+
+func (d *dumpReader) row(row []string) error {
+	if d.width != 0 && d.width != len(row) {
+		return errors.Errorf("row %d: mismatched row size %d, previously %d: %#v", d.rows+1, len(row), d.width, row)
+	}
+	d.width = len(row)
+	if err := d.f(row); err != nil {
+		return err
+	}
+	d.rows++
+
+	return nil
+}
+
+func (d *dumpReader) Process(in io.Reader) error {
+	start := 0
+	buf := make([]byte, 1024*d.bufSizeKb)
+
+	var row []string
+	var field []rune
+	var nextLiteral bool
+	var readingField bool
+
+	for {
+		n, err := in.Read(buf[start:])
+		total := n + start
+		start = 0
+
+		for i := 0; i < total; {
+			c, w := utf8.DecodeRune(buf[i:])
```

Comment:
Is it safer to use bufio.Scanner with bufio.ScanRunes? I *think* this code here can error if the Read boundary happens to be in the middle of a valid multi-byte rune.

Reply:
Added several comments describing that process.